p2p_wallet/Services/Storage/WalletMetadataService.swift

```diff
@@ -2,80 +2,122 @@
 // found in the LICENSE file.
 
+import Combine
 import Foundation
+import KeyAppKitCore
 import Onboarding
 import Resolver
 
-class WalletMetadataService: ObservableObject {
-    @Published var loading: Bool = false
-    @Published var metadata: WalletMetaData?
+actor WalletMetadataService {
+    @Injected var userWalletManager: UserWalletManager
+    @Injected var errorObserver: ErrorObserver
+
+    private let metadataSubject: CurrentValueSubject<AsyncValueState<WalletMetaData?>, Never> = .init(.init(value: nil))
+
+    nonisolated public var metadata: AsyncValueState<WalletMetaData?> {
+        metadataSubject.value
+    }
+
+    nonisolated public var metadataPublisher: AnyPublisher<AsyncValueState<WalletMetaData?>, Never> {
+        metadataSubject.eraseToAnyPublisher()
+    }
 
     private let localMetadataProvider: WalletMetadataProvider
     private let remoteMetadataProvider: WalletMetadataProvider
 
+    var subscription: [AnyCancellable] = []
+
     init(localProvider: WalletMetadataProvider, remoteProvider: WalletMetadataProvider) {
         localMetadataProvider = localProvider
         remoteMetadataProvider = remoteProvider
     }
 
-    func onboard(with metadata: WalletMetaData) async throws {
-        try await localMetadataProvider.save(metadata: metadata)
-        self.metadata = metadata
-
-        try await synchronize()
+    /// Indicator for availability of metadata.
+    nonisolated var isMetadataAvailable: Bool {
+        metadataSubject.value.value != nil
     }
 
+    /// Synchornize data between local storage and remote storage.
     func synchronize() async throws {
-        // Warm up with local data
-        metadata = try await localMetadataProvider.load()
-        
-        do {
-            loading = true
-            defer { loading = false }
-
-            // Load from cloud
-            let remoteMetadata = try await remoteMetadataProvider.load()
-
-            if
-                let metadata,
-                let remoteMetadata,
-                metadata != remoteMetadata
-            {
-                let mergedMetadata = try WalletMetaData.merge(lhs: metadata, rhs: remoteMetadata)
-
-                try await localMetadataProvider.save(metadata: mergedMetadata)
-                try await remoteMetadataProvider.save(metadata: mergedMetadata)
-
-                self.metadata = mergedMetadata
-            } else {
-                try await localMetadataProvider.save(metadata: remoteMetadata)
-                metadata = remoteMetadata
-            }
-        } catch {
-            print(error)
+        guard let userWallet = userWalletManager.wallet else {
+            let error = WalletMetadataService.Error.unauthorized
+            errorObserver.handleError(error)
             throw error
         }
+
+        if userWallet.ethAddress == nil {
+            metadataSubject.value.status = .ready
+            metadataSubject.value.value = nil
+        } else {
+            // Warm up with local data
```

Comment:
Breaking this to a function would be a lot easier to read, because it is too long.